aspnet/migration/http-modules.rst

```diff
@@ -1,8 +1,576 @@
+====================================
 
-|stub-icon| Migrating HTTP Modules to Middleware
-======================================================
+By `Matt Perdeck`_
 
-.. include:: /../common/stub-notice.txt
+This article shows how to migrate existing ASP.NET `HTTP modules and handlers <https://msdn.microsoft.com/en-us/library/bb398986.aspx>`_ to ASP.NET 5 :ref:`middleware <fundamentals-middleware>`.
 
-.. _issue: https://github.com/aspnet/Docs/issues/98
+.. contents:: In this article:
+  :local:
+  :depth: 1
+
+Handlers and modules revisited
+------------------------------
+Before proceeding to ASP.NET 5 middleware, let's first recap how HTTP modules and handlers work: 
+
+.. image:: http-modules/_static/moduleshandlers.png
+
+Handlers are:
+	* Classes that implement `IHttpHandler <https://msdn.microsoft.com/en-us/library/system.web.ihttphandler(v=vs.100).aspx>`_
+	* Used to handle requests with a given file name or extension, such as *.report*
+	* `Configured <https://msdn.microsoft.com/en-us/library/46c5ddfy(v=vs.100).aspx>`__ in the *web.config*
+
+Modules are:
+	* Classes that implement `IHttpModule <https://msdn.microsoft.com/en-us/library/system.web.ihttpmodule(v=vs.100).aspx>`_
+	* Invoked for every request
+	* Able to stop further processing of a request
+	* Able to add to the HTTP response, or create their own
+	* `Configured <https://msdn.microsoft.com/en-us/library/ms227673(v=vs.100).aspx>`__ in the web.config
+
+**The order in which modules process incoming requests is determined by:**
+
+	1.	The `application life cycle <https://msdn.microsoft.com/en-us/library/ms227673(v=vs.100).aspx>`_, 
+		which is a series events fired by ASP.NET
+		- 
+		`BeginRequest <https://msdn.microsoft.com/en-us/library/system.web.httpapplication.beginrequest(v=vs.100).aspx>`_,
+		`AuthenticateRequest <https://msdn.microsoft.com/en-us/library/system.web.httpapplication.authenticaterequest(v=vs.100).aspx>`_, etc.
+		Each module can create a handler for one or more events.
+
+	2. When handling the same event, the order in which they are configured in the web.config.
+
+In addition to modules, you can add handlers for the life cycle events to your *Global.asax.cs* file. These handlers run after the handlers in the configured modules.
+
+From handlers and modules to middleware
+---------------------------------------
+Middleware is simpler than HTTP modules and handlers:
+    * Modules, handlers, Global.asax.cs, web.config (except for IIS configuration) and the application life cycle are gone
+    * The roles of both modules and handlers have been taken over by middleware
+    * Middleware is configured using code instead of in the web.config
+    * :ref:`Pipeline branching <middleware-run-map-use>` lets you send requests to specific middleware, based on not only the URL but also on request headers, etc.
+
+Middleware is very similar to modules:
+	* Invoked in principle for every request
+	* Able to stop further processing of a request
```

Comment:
Able to short-circuit a request. (Each delegate has the opportunity to perform operations before and after the next delegate. Any delegate can choose to stop passing the request on to the next delegate, and instead handle the request itself. For more information see fundamentals/middleware)